File: src/lib/wow/server/runtime.ts
/**
 * Minimal runtime state replacing wow.export's core.js couplings
 * (load function registry, selected CDN region, progress shim).
 */
import type { CASC } from '@/lib/wow/archive/casc/casc-source';
import { write } from '@/lib/wow/log';

export type LoadFunc = () => Promise<void> | void;

const loadFuncs: LoadFunc[] = [];

/**
 * Register a function to be executed when a CASC source finishes loading
 * (mirrors core.registerLoadFunc; used by DB caches).
 */
export function registerLoadFunc(fn: LoadFunc): void {
  loadFuncs.push(fn);
}

/** Run all registered load functions (mirrors core.runLoadFuncs). */
export async function runLoadFuncs(): Promise<void> {
  for (const fn of loadFuncs) await fn();
}

/** Selected CDN region tag, used for CDN fallback of local installations. */
export const runtimeState: {
  selectedCDNRegionTag: string;
  /** Active CASC source (replaces wow.export's core.view.casc); set once loaded. */
  casc: CASC | null;
} = {
  selectedCDNRegionTag: 'us',
  casc: null,
};

/** Get the active CASC source, throwing if none is loaded. */
export function getCasc(): CASC {
  if (!runtimeState.casc) throw new Error('No CASC source has been loaded.');
  return runtimeState.casc;
}

/** Progress shim replacing core.createProgress(); just logs steps. */
export interface Progress {
  step(name?: string): Promise<void>;
}

export function createProgress(_segments?: number): Progress {
  return {
    step(name?: string): Promise<void> {
      if (name) write('Progress: %s', name);
      return Promise.resolve();
    },
  };
}

File: golang/internal/wow/runtime/runtime.go
// Package runtime provides minimal runtime state for the WoW reader.
package runtime

import "github.com/pqhuy98/wow-converter/internal/wow/log"

// LoadFunc is executed when a CASC source finishes loading.
type LoadFunc func() error

var loadFuncs []LoadFunc

// RegisterLoadFunc registers a load function.
func RegisterLoadFunc(fn LoadFunc) {
	loadFuncs = append(loadFuncs, fn)
}

// RunLoadFuncs runs all registered load functions.
func RunLoadFuncs() error {
	for _, fn := range loadFuncs {
		if err := fn(); err != nil {
			return err
		}
	}
	return nil
}

// RuntimeState holds selected CDN region and active CASC source.
var RuntimeState = struct {
	SelectedCDNRegionTag string
	Casc                 any
}{
	SelectedCDNRegionTag: "us",
}

// Progress is a progress shim replacing core.createProgress().
type Progress interface {
	Step(name string) error
}

type progress struct{}

// CreateProgress creates a progress tracker.
func CreateProgress(_ int) Progress {
	return progress{}
}

func (progress) Step(name string) error {
	if name != "" {
		log.Write("Progress: %s", name)
	}
	return nil
}
